### PR TITLE
Install Git LFS

### LINKED ISSUE #3: Add Git LFS support
The current Flatpak does not have Git LFS binaries or filters installed.

It might be possible to reuse code from [com.axosoft.GitKraken](https://github.com/flathub/com.axosoft.GitKraken), [com.syntevo.SmartGit](https://github.com/flathub/com.syntevo.SmartGit) or [org.sparkleshare.SparkleShare](https://github.com/flathub/org.sparkleshare.SparkleShare).

### PR DESCRIPTION
**Pull request by fnkr**
Fixes #3.

### CODE REVIEW

**Comment by fnkr** [comment]
@ptomato Can you please take a look at this?

**Comment by ptomato** [comment]
Sorry, it's on my to-do list but I'm kind of busy this week.

File: com.sublimemerge.App.yaml
Comment (ptomato) [inline com.sublimemerge.App.yaml]: I would suggest moving the `git-lfs` module earlier in the list than the `sublime_merge` module, so that when we update sublime merge we can still use the cached build of git-lfs.

File: com.sublimemerge.App.yaml
Comment (ptomato) [inline com.sublimemerge.App.yaml]: You can probably just put this line directly in the `sublime_merge.sh` script where it's created above (currently lines 41-44)

File: com.sublimemerge.App.yaml
Comment (ptomato) [inline com.sublimemerge.App.yaml]: I would suggest putting this in the `sublime_merge.sh` script below as well, so that you can append to the path, e.g. `export PATH="$PATH:/app/libexec/git-lfs"`. I think that way it will still work if someone tries to override `--env=PATH` on the command line.

**Review by ptomato — CHANGES_REQUESTED** [review]
I have a couple of suggestions to make the manifest simpler, but otherwise this looks great, I'll be happy to merge it once the changes are made. I can also just make these changes myself if you don't have time right now. I will rely on you to test the functionality thoroughly though, as I have never used git-lfs :smile:

**Comment by fnkr** [comment]
I updated the pull request to match your recommendations and tested it with no overrides and with `flatpak override com.sublimemerge.App --nofilesystem=host --filesystem='~/Workspace'` too and it works.

**Comment by ptomato** [comment]
Thanks!

**Comment by ptomato** [comment]
Hmm, after installing the test build I can no longer use the command-line git client on the repository that I tested it with... (that is, the host system git client, outside the Flatpak sandbox, where I don't have git-lfs) 

> This repository is configured for Git LFS but 'git-lfs' was not found on your path. If you no longer wish to use Git LFS, remove this hook by deleting .git/hooks/post-checkout.

Once I delete that file the repository works with the host system client again.

@fnkr Do you know if it's possible to stop Sublime Merge from reconfiguring the repository unless you actually require git-lfs use?

**Comment by fnkr** [comment]
LFS writes to Git's `core.hooksPath`. We could override that and use something within the sandbox instead.

```sh
git config --global core.hooksPath /etc/git/hooks
```

The command could be added to `sublime_merge.sh`. Git will create the directory if it doesn't exist.

**Comment by ptomato** [comment]
If I'm understanding the documentation correctly, that would only affect new repositories? What happened with me is that an existing repository got the hooks added to it.

**Comment by fnkr** [comment]
It will not fix your existing repository but it will prevent LFS from adding hooks to the local hook directory of any new or existing repository. Git and LFS will use `/etc/git/hooks` (which is sandboxed) instead of `.git/hooks` (relative to your Git workdir) and therefore Git operations from outside of the sandbox won't break anymore.

**Comment by ptomato** [comment]
I see, but wouldn't that ignore any other hooks that you did have in `.git/hooks`, like linter checks?

**Comment by fnkr** [comment]
Unfortunately yes. Git does not support to have multiple hooks. On the other hand, such hooks would probably fail anyway because the necessary tools (e.g. pylint) would not be available from within the sandbox. And it would still be possible to use local hooks again. LFS won't override existing Git hooks.

```sh
# Change the repository-specific Git config to use local hooks again
git config --local core.hooksPath .git/hooks
```

Another solution would be not to install LFS filters and hooks by default and leave this up to the user. If the following line is removed the Flatpak would only provide the LFS binary but not use it by default:

https://github.com/flathub/com.sublimemerge.App/blob/cfe896ceb6cfd28ac3d0ff4d3f7472fdde8b38e1/com.sublimemerge.App.yaml#L55

The user could then enable LFS like this:

```sh
$ mkdir -p ~/.var/app/com.sublimemerge.App/config/git
$ tee -a ~/.var/app/com.sublimemerge.App/config/git/config <<EOF
[filter "lfs"]
    required = true
    clean = git-lfs clean -- %f
    smudge = git-lfs smudge -- %f
    process = git-lfs filter-process
EOF
```

**Comment by ptomato** [comment]
I do have Sublime Merge set up to use my local hooks — since it basically has host filesystem access, it can look in the host's path for clang-format, for example.

I do find it a bit odd that it's updating the hooks in a repo where git-lfs hasn't been used yet. Do you know if this is git-lfs doing that, or is it that Sublime Merge sees that git-lfs is installed and figures that since you have it on your system, it's harmless to install the hooks? In the latter case maybe we could report the issue to Sublime Merge.

I wonder if it would be feasible to add git-lfs as an [extension](https://github.com/flatpak/flatpak/wiki/Extensions) named something like `com.sublimemerge.App.GitLfsUtils`; we already do something like this for Sublime Text and common dev/linter packages. That way only people who were intending to use git-lfs would download the extension.

**Comment by fnkr** [comment]
It's the LFS filter that auto-installs LFS hooks. I was not able to find any option to disable this behaviour. I would not do that anyway because if filters were active but hooks were not, Git would add LFS objects to the LFS storage area but never push them, which could cause data loss.

I think creating an extension for LFS is a great idea. We could provide the LFS binary for all users in the standard Flatpak and create an optional extension which would enable LFS by default, so the user can choose:

- Use the Flatpak as is and enable LFS per repository if needed (just run `git lfs install --local` on the host and LFS will work for that repository on the command line and in Sublime Merge too)
- Install the extension to enable LFS globally, with the downside being that local Git hooks would be disabled (can be reactivated per repository with `git config --local core.hooksPath .git/hooks`)

What do you think about that?

**Comment by ptomato** [comment]
That sounds mostly good to me. I wonder if it would be possible to do it without changing the hooks config path, so that when you install the extension you opt-in to whatever the normal behaviour would be if you installed git-lfs on the host?

**Comment by fnkr** [comment]
That would work as long as LFS is installed on the host too.

How do we proceed? First step would probably be to remove this line:

https://github.com/flathub/com.sublimemerge.App/blob/cfe896ceb6cfd28ac3d0ff4d3f7472fdde8b38e1/com.sublimemerge.App.yaml#L55

Second step would be to create an extension to run the command every time the application starts.

**Comment by fnkr** [comment]
@ptomato Shall I open a merge request for the first step?

**Comment by fnkr** [comment]
@ptomato I've created a pull request for the first step (see #5). This will fix the master branch and we can create an LFS extension afterwards.

**Pull request merged**
